README.md
<!--- Local IspellDict: en -->
<!--- SPDX-FileCopyrightText: 2017-2022 Jens Lechtenbörger -->
<!--- SPDX-License-Identifier: GPL-3.0-or-later -->

[![DOI](https://jose.theoj.org/papers/10.21105/jose.00050/status.svg)](https://doi.org/10.21105/jose.00050)
[![REUSE status](https://api.reuse.software/badge/gitlab.com/oer/emacs-reveal)](https://api.reuse.software/info/gitlab.com/oer/emacs-reveal)

This repository provides *emacs-reveal*, which is
[free/libre and open source software (FLOSS)](https://en.wikipedia.org/wiki/Free_and_open-source_software)
to create [Open Educational Resources (OER)](https://en.wikipedia.org/wiki/Open_educational_resources)
in the form of [reveal.js](https://revealjs.com/) presentations (HTML
slides with audio).  As the name suggests, emacs-reveal is based on
the text editor [GNU Emacs](https://www.gnu.org/software/emacs/).
More specifically, authors write their OER in a lightweight markup
language called [Org mode](https://orgmode.org/), which is native to
GNU Emacs but can be edited everywhere (as text file; also with
“real” support in [other editors](https://orgmode.org/install.html#editors))
and which allows authors (a) to
collaborate on source documents via version control systems such as
Git and (b) to focus on contents, while layout information is
maintained separately.  Org files can then be exported to various
target formats (e.g., reveal.js presentations or PDF documents).

<!-- START doctoc generated TOC please keep comment here to allow auto update -->
<!-- DON'T EDIT THIS SECTION, INSTEAD RE-RUN doctoc TO UPDATE -->
**Table of Contents**  *generated with [DocToc](https://github.com/thlorenz/doctoc)*

- [Relationships between Emacs packages](#relationships-between-emacs-packages)
- [Introduction](#introduction)
- [Technical Background](#technical-background)
- [Publications on emacs-reveal](#publications-on-emacs-reveal)
- [Usage of emacs-reveal](#usage-of-emacs-reveal)
  - [Emacs-reveal in GitLab CI/CD](#emacs-reveal-in-gitlab-cicd)
  - [Emacs-reveal with Docker](#emacs-reveal-with-docker)
  - [Emacs-reveal for daily use](#emacs-reveal-for-daily-use)
- [Emacs-reveal for a course on Operating Systems](#emacs-reveal-for-a-course-on-operating-systems)
- [Evaluation results](#evaluation-results)

<!-- END doctoc generated TOC please keep comment here to allow auto update -->

# Relationships between Emacs packages

| Package                                                                 | Description                                                                                                                                      |
|-------------------------------------------------------------------------|--------------------------------------------------------------------------------------------------------------------------------------------------|
| [org-reveal](https://github.com/yjwen/org-reveal)                       | Origin of org-re-reveal                                                                                                                          |
| [org-re-reveal](https://gitlab.com/oer/org-re-reveal)                   | Fork of org-reveal, initially to add audio fragments, now with [various changes](https://gitlab.com/oer/org-re-reveal/blob/master/CHANGELOG.org) |
| [org-re-reveal-ref](https://gitlab.com/oer/org-re-reveal-ref)           | Addon to org-re-reveal for bibliography slide based on [org-ref](https://github.com/jkitchin/org-ref)                                            |
| [org-re-reveal-citeproc](https://gitlab.com/oer/org-re-reveal-citeproc) | Alternative to org-re-reveal-ref for bibliography slide based on syntax of Org mode 9.5                                                          |
| [oer-reveal](https://gitlab.com/oer/oer-reveal)                         | Export backend derived from org-re-reveal; functionality for installation of reveal.js and plugins; simplification of licensing for OER          |
| [emacs-reveal](https://gitlab.com/oer/emacs-reveal/)                    | Bundling of org-re-reveal, org-re-reveal-citeproc, org-re-reveal-ref, and oer-reveal                                                             |

Originally, *emacs-reveal* was created to enhance *org-reveal*, and it
contained the code of what is now maintained separately as
*org-re-reveal-ref* and *oer-reveal*.  Those packages were separated
to make as much of *emacs-reveal* available on the Emacs package
repository [MELPA](https://melpa.org/) as possible.
Whether anyone wants to use those package in isolation is up to them.

# Introduction

During summer term 2017 I created emacs-reveal to
produce lecture slides (initially for a course on Operating Systems)
and presentations as OER under the
[Creative Commons license](https://creativecommons.org/use-remix/cc-licenses/)
[CC BY-SA 4.0](https://creativecommons.org/licenses/by-sa/4.0/)
using the
HTML presentation framework [reveal.js](https://github.com/hakimel/reveal.js/)
(see [here for my presentations on Operating Systems](https://oer.gitlab.io/OS/)).

Since its creation, I’ve been updating emacs-reveal and the
surrounding infrastructure continuously.  In particular, sharing of
OER figures with proper attribution and license information has been
simplified considerably; see the project hosting
[OER figures](https://gitlab.com/oer/figures/) and
[this paper](https://gitlab.com/lechten/publications/blob/master/delfi2019/delfi.org)
([published version](https://dx.doi.org/10.18420/delfi2019_280))
for more details on my approach towards meta-data, which is based on
standard vocabularies and embeds machine-readable license information using
the [Creative Commons Rights Expression Language (CC REL)](https://labs.creativecommons.org/2011/ccrel-guide/)
(with [RDFa](https://wiki.creativecommons.org/wiki/RDFa) in HTML),
(a) avoiding manual identification and copying of licensing
information and (b) making licensing information accessible on the
[Semantic Web](https://en.wikipedia.org/wiki/Semantic_Web).

Beyond that paper’s contents, emacs-reveal supports the specification
of a presentation’s primary license with SPDX headers (see the
[REUSE project](https://reuse.software/) for more information on such
headers), from which license information can be generated (“Except
where otherwise noted, the work x by y is published under the license
z”), e.g.,
[that slide](https://oer.gitlab.io/emacs-reveal-howto/howto.html#slide-license),
which includes machine-readable RDFa markup (the
[PDF version of that presentation](https://oer.gitlab.io/emacs-reveal-howto/howto.pdf)
just contains a human readable variant).

I’m using emacs-reveal for my courses and talks in general.  Feel free
to use and adapt for your purposes!

# Technical Background

Emacs-reveal addresses several technical requirements for OER
(explained in detail in
[this paper (in German)](https://www.medienpaed.com/article/view/651)):
1. Layout and style of presentations are separated from content to
   simplify *collaboration across organizational boundaries*.
2. Slides’ content is written down in simple text files, which enables
   *comparisons* of adapted or enhanced *versions* (with `diff`-like
   functionality, e.g, with the
   [compare functionality on GitLab](https://gitlab.com/oer/OS/compare/os02...os03)).
3. Presentations are usable in standard Web browsers with their
   multimedia support on almost *any device*, without the need for
   special-purpose software.  Offline use is supported.

Although reveal.js is an HTML presentation framework, I do not create
presentations in HTML.  Instead, I rely on my usual computing
environment, namely the text editor
[GNU Emacs](https://www.gnu.org/software/emacs/), in a simple
text format called [Org mode](https://orgmode.org/), which can be
exported to reveal.js thanks to
[Org-Reveal](https://github.com/yjwen/org-reveal/), for which I
maintain [the forked variant org-re-reveal](https://gitlab.com/oer/org-re-reveal),
its extensions [org-re-reveal-ref](https://gitlab.com/oer/org-re-reveal-ref)
and [org-re-reveal-citeproc](https://gitlab.com/oer/org-re-reveal-citeproc)
for citations and bibliography, and
[oer-reveal](https://gitlab.com/oer/oer-reveal) with further OER resources.
That way, I can focus on slides’ contents in a highly intuitive plain
text document, which can be exported to HTML for presentation in a Web
browser (besides, presentations can be downloaded in two different
layouts in PDF format in response to students’ requests:
one PDF version contains one page per slide while the other one is
a more concise version generated via [LaTeX](https://www.latex-project.org/)
from Org source documents).  All layout
information is kept separately from content, so little effort is
necessary to adapt presentations to different styles or corporate
identities.

My requirements, which lead me to select reveal.js, are as follows:
 1. Free software (really, I mean
    [free](https://fsfe.org/about/basics/freesoftware.en.html), not
    gratis) that is usable on lots of devices
 2. Embed audio recordings for self-contained presentations
 3. Focus on content, not on layout while creating presentations
 4. Allow collaboration, ideally with support for versions and diffs

First, I tried
[LaTeX with the Beamer package](https://en.wikibooks.org/wiki/LaTeX/Presentations),
where presentations are created as PDF documents and
which seems to meet all requirements.  However, although audio can be
embedded, I found audio support in free PDF readers unsatisfactory.

Next, I tried [LibreOffice Impress](https://www.libreoffice.org/).
I perceived the creation of presentations as too cumbersome, in
particular embedding of audio files.  Also, Impress does neither meet
requirements (3) nor (4).

Eventually, I ended up with reveal.js.
To embed audio, I’m using the
[audio-slideshow plugin of reveal.js](https://github.com/rajgoel/reveal.js-plugins).

The setup is as follows:
 * GNU Emacs with the following packages
   * Org mode (recent versions, e.g., from ELPA)
   * Org-re-reveal, org-re-reveal-citeproc, org-re-reveal-ref, oer-reveal
     (from MELPA or [my repositories](https://gitlab.com/oer))
   * Htmlize (for source code highlighting)
 * reveal.js
   * [reveal-a11y](https://github.com/marcysutton/reveal-a11y)
     (for improved accessibility)
   * [reveal.js-plugins](https://github.com/rajgoel/reveal.js-plugins)
     (for audio explanations and SVG animations)
   * [reveal.js-coursemod](https://github.com/Sonaryr/reveal.js-coursemod)
     (for a courseware view where notes are visible)
   * [reveal.js-jump-plugin](https://github.com/SethosII/reveal.js-jump-plugin)
     (for jumps to slides by entering their numbers)
   * [reveal.js-quiz](https://gitlab.com/schaepermeier/reveal.js-quiz)
     (for embedded quizzes supporting retrieval practice)
   * [Reveal.js-TOC-Progress](https://github.com/e-gor/Reveal.js-TOC-Progress)
     (for a hyperlinked table of contents as slide footer)
 * LaTeX for PDF generation

# Publications on emacs-reveal

* [Emacs-reveal: A software bundle to create OER presentations](https://doi.org/10.21105/jose.00050)
* [Erstellung und Weiterentwicklung von Open Educational Resources im Selbstversuch](https://www.medienpaed.com/article/view/651)
* [Simplifying license attribution for OER with emacs-reveal](https://dx.doi.org/10.18420/delfi2019_280)

# Usage of emacs-reveal

To use emacs-reveal, several options exist as sketched subsequently.

1. Use emacs-reveal inside [GitLab’s Continuous Integration](https://docs.gitlab.com/ce/ci/)
   infrastructure with automatic deployment of resulting presentations
   as [GitLab Pages](https://docs.gitlab.com/ce/user/project/pages/index.html).
2. Use Docker image `registry.gitlab.com/oer/emacs-reveal/emacs-reveal` from
   [this container registry](https://gitlab.com/oer/emacs-reveal/container_registry)
3. Install GNU Emacs, emacs-reveal, and above packages.

## Emacs-reveal in GitLab CI/CD

A simple way to get started with the first option is to fork the
[emacs-reveal howto](https://gitlab.com/oer/emacs-reveal-howto) at
GitLab.  In your clone, change the file `howto.org` and commit the
change.  A GitLab runner will start automatically and publish the
presentation under your
[GitLab Pages namespace](https://docs.gitlab.com/ce/user/project/pages/getting_started_part_one.html).

## Emacs-reveal with Docker

For the second option, [install Docker](https://docs.docker.com/install/)
and start a container for emacs-reveal, e.g.:

```
docker run --rm --name emacs-reveal -it registry.gitlab.com/oer/emacs-reveal/emacs-reveal
```

For GUI support, you need to set up authentication information for the
Docker container.  The details depend on your environment.  If you
have got an environment variable `XAUTHORITY` in your host system, the
following should work:

```
docker run --rm --name emacs-reveal -it --net=host -e DISPLAY -v $XAUTHORITY:/root/.Xauthority registry.gitlab.com/oer/emacs-reveal/emacs-reveal
```

Alternatively, if you have got a file `~/.Xauthority`:

```
docker run --rm --name emacs-reveal -it --net=host -e DISPLAY -v $HOME/.Xauthority:/root/.Xauthority registry.gitlab.com/oer/emacs-reveal/emacs-reveal
```

In any case, a shell opens in the container.  Type `emacs` to start
GNU Emacs, load an `org` file (e.g., `Readme.org` coming with
`org-re-reveal`, available under directory
`/root/.emacs.d/elpa/emacs-reveal/org-re-reveal`)
and export the reveal.js presentation (with function
`oer-reveal-export-to-html`, bound to key `C-c C-e w w`).

Copy the generated HTML presentation (and resources `local.css` as well as
`images/`) from the container to a directory with your `reveal.js`
installation.  E.g., use commands such as the following:

```

CHANGELOG.org
# Local IspellDict: en
# SPDX-License-Identifier: GPL-3.0-or-later
# SPDX-FileCopyrightText: 2020-2022 Jens Lechtenbörger

* Changelog
This file is based on
[[https://keepachangelog.com/en/1.0.0/][Keep a Changelog]],
where version numbers starting with version 4.0.0 adhere to
[[https://semver.org/spec/v2.0.0.html][Semantic Versioning]].

** [Unreleased]

** [9.16.3] - 2022-12-29
*** Fixed
    - Update oer-reveal to 4.10.2

** [9.16.2] - 2022-12-29
*** Fixed
    - Update oer-reveal to 4.10.1

** [9.16.0],[9.16.1] - 2022-12-28
*** Changed
    - Update org-re-reveal to 3.17.0
    - Update oer-reveal to 4.10.0
    - These changes allow to configure CSS with plugins directly,
      related to issue #35
    - Version 9.16.0 tagged as 9.16.1 by accident

** [9.15.1] - 2022-11-21
*** Fixed
    - Update oer-reveal to 4.9.1

** [9.15.0] - 2022-11-20
*** Changed
    - Update oer-reveal to 4.9.0

** [9.14.1] - 2022-10-07
*** Fixed
    - Update org-re-reveal to 3.16.1

** [9.14.0] - 2022-10-05
*** Changed
    - Update oer-reveal to 4.8.0

** [9.13.0] - 2022-10-05
*** Changed
    - Update org-re-reveal to 3.16.0

** [9.12.0] - 2022-08-04
*** Changed
    - Update oer-reveal to 4.7.0

** [9.11.0] - 2022-08-02
*** Changed
    - Update oer-reveal to 4.6.0

** [9.10.0] - 2022-06-13
*** Added
    - URL parameter ~audio-speed~
      - Update emacs-reveal-submodules to 2.4.0 with own branch
        oer-audio (with option to control audio speed)
      - Update oer-reveal to 4.5.0

** [9.9.5] - 2022-06-03
*** Fixed
    - Update emacs-reveal-submodules to 2.3.3 with own branch of
      reveal.js-plugins (temporarily until issue #147 is fixed)
    - Update oer-reveal to 4.4.4

** [9.9.4] - 2022-06-03
*** Fixed
    - Update emacs-reveal-submodules to 2.3.2 with downgrade of
      Anything plugin
      (https://github.com/rajgoel/reveal.js-plugins/issues/147)
    - Update oer-reveal to 4.4.3

** [9.9.3] - 2022-05-16
*** Changed
    - Update oer-reveal to 4.4.2

** [9.9.2] - 2022-05-16
*** Fixed
    - Update oer-reveal to 4.4.1

** [9.9.1] - 2022-05-14
*** Fixed
    - Restrict build context for image selenium-chrome

** [9.9.0] - 2022-05-14
*** Changed
    - Move updated selenium-print-pdf.py to
      [[file:docker/code/selenium-print-pdf.py]]

** [9.8.2] - 2022-05-13
*** Changed
    - Increase wait time in [[file:python/selenium-print-pdf.py]]

** [9.8.1] - 2022-05-13
*** Fixed
    - Use file URIs in [[file:python/selenium-print-pdf.py]]

** [9.8.0] - 2022-05-13
*** Added
    - Docker image for Selenium with Chrome and Python bindings

** [9.7.0] - 2022-05-13
*** Added
    - Script [[file:python/selenium-print-pdf.py]] to print
      presentation to PDF with Selenium

** [9.6.0] - 2022-04-02
*** Changed
    - Update org-re-reveal to 3.13.0 and oer-reveal to 4.4.0

** [9.5.1] - 2022-04-02
*** Fixed
    - Update oer-reveal to 4.3.1

** [9.5.0] - 2022-04-02
*** Changed
    - Update emacs-reveal-submodules to 2.3.1, with reveal.js 4.3.1
      and fork of reveal.js-quiz

** [9.4.1] - 2022-03-22
*** Fixed
    - Revert update of emacs-reveal-submodules to 2.3.0,
      breaks quiz plugin with reveal.js 4.3.1

** [9.4.0] - 2022-03-22
*** Changed
    - Update emacs-reveal-submodules to 2.3.0, with reveal.js 4.3.1

** [9.3.0] - 2022-03-21
*** Changed
    - Update oer-reveal to 4.3.0

** [9.2.1] - 2022-03-21
*** Fixed
    - Update oer-reveal to 4.2.1

** [9.2.0] - 2022-03-20
*** Changed
    - Update oer-reveal to 4.2.0

** [9.1.2] - 2022-01-17
*** Fixed
    - Update oer-reveal to 4.1.2; addresses issue #30

** [9.1.1] - 2021-12-07
*** Fixed
    - Update emacs-reveal-submodules to 2.2.3; addresses issue #28

** [9.1.0] - 2021-10-29
*** Changed
    - Update oer-reveal to 4.1.0

** [9.0.1] - 2021-10-28
*** Fixed
    - Initialization for org-re-reveal-citeproc

** [9.0.0] - 2021-10-28
*** Added
    - Submodule org-re-reveal-citeproc to support new citation links
      of Org mode 9.5
    - Variable [[file:emacs-reveal.el::(defcustom emacs-reveal-cite-pkg][emacs-reveal-cite-pkg]]
      to choose between org-re-reveal-ref and org-re-reveal-citeproc
*** Changed
    - Switch to oer-reveal 4.0.0.  In particular,
      oer-reveal-publish-setq-defaults now changes fewer parts of the
      user’s environment.

** [8.30.1] - 2021-10-26
*** Fixed
    - Org-ref 3 introduced
      [[https://github.com/jkitchin/org-ref#breaking-changes-in-version-3][breaking changes]].
      Switch to MELPA stable to install version 2.

** [8.30.0] - 2021-10-15
*** Changed
    - Update oer-reveal to 3.25.0

** [8.29.0] - 2021-10-02
*** Changed
    - Update Org mode to 9.5
    - Update submodules to upstream versions

** [8.28.0] - 2021-09-15
*** Changed
    - Update oer-reveal to 3.24.0

** [8.27.0] - 2021-08-27
*** Changed
    - Update oer-reveal to 3.23.0

** [8.26.0] - 2021-08-26
*** Changed
    - Update oer-reveal to 3.22.0

** [8.25.3] - 2021-08-19
*** Fixed
    - Update oer-reveal to 3.21.2

** [8.25.2] - 2021-08-19
*** Fixed
    - Update oer-reveal to 3.21.1

** [8.25.1] - 2021-08-17
*** Fixed
    - Issue #27 related to testing

** [8.25.0] - 2021-08-17
*** Changed
    - Update oer-reveal to 3.21.0
    - Update org-re-reveal to 3.10.0

** [8.24.2] - 2021-06-23
*** Fixed
    - Update oer-reveal to 3.20.2

** [8.24.1] - 2021-05-21
*** Fixed
    - Update oer-reveal to 3.20.1

** [8.24.0] - 2021-05-12
*** Changed
    - Update oer-reveal to 3.20.0

** [8.23.0] - 2021-05-12
*** Changed
    - Update oer-reveal to 3.19.0
    - Update org-re-reveal to 3.9.0

** [8.22.3] - 2021-04-18
*** Fixed
    - Update oer-reveal to 3.18.3

** [8.22.2] - 2021-04-17
*** Fixed
    - Update oer-reveal to 3.18.2

** [8.22.1] - 2021-04-17
*** Fixed
    - Update oer-reveal to 3.18.1

** [8.22.0] - 2021-04-17
*** Changed
    - Update oer-reveal to 3.18.0

** [8.21.2] - 2021-04-05
*** Fixed
    - Update org-re-reveal to 3.8.1

** [8.21.1] - 2021-04-05
*** Fixed
    - Update oer-reveal to 3.17.1

** [8.21.0] - 2021-04-05
*** Changed
    - Update org-re-reveal to 3.8.0


emacs-reveal.el
;;; emacs-reveal.el --- Create OER as reveal.js presentations with Emacs  -*- lexical-binding: t; -*-
;; -*- Mode: Emacs-Lisp -*-
;; -*- coding: utf-8 -*-

;; SPDX-FileCopyrightText: 2017-2022 Jens Lechtenbörger
;; SPDX-License-Identifier: GPL-3.0-or-later

;; Author: Jens Lechtenbörger
;; URL: https://gitlab.com/oer/emacs-reveal
;; Version: 9.16.3
;; Package-Requires: ((emacs "25.1") (oer-reveal "3.0.0") (org-re-reveal-citeproc "2.0.1") (org-re-reveal-ref "1.0.0"))
;; Keywords: hypermedia, tools, slideshow, presentation, OER

;; Emacs-reveal supports citations that are hyperlinked to a bibliography
;; slide.  Traditionally, org-ref supported citations for Org mode, and
;; org-re-reveal-ref integrated org-ref into emacs-reveal.
;; In version 9.5, Org mode added a new citation format, which is supported
;; by citeproc, which is integrated with org-re-reveal-citeproc.
;; Customizable variable emacs-reveal-cite-pkg determines whether
;; org-re-reveal-ref or org-re-reveal-citeproc is used.
;; If dependencies (org-ref or citeproc) are missing,
;; installation from MELPA is offered.

;;; License:
;; This program is free software; you can redistribute it and/or
;; modify it under the terms of the GNU General Public License as
;; published by the Free Software Foundation; either version 3, or
;; (at your option) any later version.

;; This program is distributed in the hope that it will be useful,
;; but WITHOUT ANY WARRANTY; without even the implied warranty of
;; MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
;; General Public License for more details.

;; You should have received a copy of the GNU General Public License
;; along with GNU Emacs; see the file COPYING.
;; If not, see http://www.gnu.org/licenses/ or write to the
;; Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor,
;; Boston, MA 02110-1301, USA.

;;; Commentary:
;; The software emacs-reveal helps to create HTML presentations (with
;; audio explanations if you wish) with reveal.js from Org mode files
;; in GNU Emacs with org-re-reveal (a fork of org-reveal) and several
;; reveal.js plugins.  Generated presentations are usable with standard
;; browsers, also mobile and offline.
;;
;; Emacs-reveal grew out of a forked version of org-reveal
;; (https://github.com/yjwen/org-reveal) when its development stopped.
;; This led to the creation of org-re-reveal, org-re-reveal-ref, and
;; oer-reveal, upon which emacs-reveal is built.
;;
;; Just as org-reveal, emacs-reveal provides an export back-end for Org
;; mode (see https://orgmode.org/manual/Exporting.html).  As such, it
;; comes with Org's separation of contents and layout, allowing to
;; create presentations in a fashion similarly to Beamer LaTeX,
;; including the use of BibTeX files for bibliographic notes (for
;; classroom presentations), hyperlinks within and between
;; presentations, and the generation of a keyword index.
;; Beyond other similar projects, emacs-reveal comes with mechanisms
;; (a) to add audio explanations to presentations and (b) to share
;; free and open images and figures with proper attribution
;; information for their inclusion in open educational resources
;; (OER).  See there for OER presentations (HTML with reveal.js and
;; PDF generated from Org files via LaTeX) for a university course on
;; Operating Systems that are generated with emacs-reveal:
;; https://oer.gitlab.io/OS/
;;
;; A howto for the use of emacs-reveal is available over there:
;; https://gitlab.com/oer/emacs-reveal-howto/blob/master/howto.org
;; https://oer.gitlab.io/emacs-reveal-howto/howto.html (generated HTML)
;;
;; * Installation
;; You need Git to use emacs-reveal.  See there:
;; https://git-scm.com/book/en/v2/Getting-Started-Installing-Git
;;
;; Emacs-reveal is only available from GitLab.
;;
;; By default, reveal.js presentations and a PDF variant are generated,
;; which requires a LaTeX installation.
;; See `oer-reveal-publish-org-publishing-functions'.
;;
;; The howto mentions several alternatives to install emacs-reveal.
;; Here is one:
;; 1. Choose directory, e.g., ~/.emacs.d/elpa, and clone software
;;    - cd ~/.emacs.d/elpa
;;    - git clone --recursive https://gitlab.com/oer/emacs-reveal.git
;;      - (Option `--recursive' downloads submodules)
;; 2. Add following lines to ~/.emacs
;;    - (add-to-list 'load-path "~/.emacs.d/elpa/emacs-reveal")
;;    - (require 'emacs-reveal)
;; 3. Restart Emacs (installation of `org-ref' or `citeproc' is offered,
;;    if necessary)
;;
;; * Usage
;; Load an Org file (e.g., Readme.org coming with `org-re-reveal') and
;; export it to HTML (e.g., `C-c C-e w w').  Note that this just
;; generates the HTML file but does neither publish images nor other
;; resources such as JavaScript or CSS files.
;; To publish a project with all related resources, please check out
;; the emacs-reveal howto mentioned above.
;; In particular, the howto contains a sample file "elisp/publish.el" to
;; publish reveal.js presentations from Org source files.
;; Note that the HTML version of the howto is generated in a Docker
;; image on GitLab; its YML configuration shows the necessary steps
;; to generate and publish the project:
;; https://gitlab.com/oer/emacs-reveal-howto/blob/master/.gitlab-ci.yml
;;
;; With version 4, reveal.js introduced incompatible changes in some paths.
;; The target version of reveal.js to be used with emacs-reveal is
;; determined by oer-reveal-revealjs-version, which in turn takes
;; precedence over org-re-reveal-revealjs-version.  Please consult its
;; doc string.

;;; Code:
(defgroup org-export-emacs-reveal nil
  "Options for exporting Org files to reveal.js HTML pressentations.
The options here are provided by package `emacs-reveal'.  They extend those
of `oer-reveal'."
  :tag "Org Export emacs-reveal"
  :group 'org-export-oer-reveal)

(defcustom emacs-reveal-managed-install-p t
  "Configure whether to update `emacs-reveal' and submodules or not.
By default, `emacs-reveal' tries to update itself and its submodules via Git.
This requires `git' and `make' to be installed.
If you set this to nil, you should install Lisp packages in
`emacs-reveal-lisp-packages' yourself.  In that case, you also need to
decide how to go about submodules of emacs-reveal, which are managed by
`oer-reveal' by default (maybe change `oer-reveal-submodules-dir' to
point to your manually managed directory and set
`oer-reveal-submodules-version' to nil)."
  :group 'org-export-emacs-reveal
  :type 'boolean
  :package-version '(emacs-reveal . "7.0.0"))

(defvar emacs-reveal-install-dir
  (file-name-directory (or load-file-name (buffer-file-name)))
  "Directory of file \"emacs-reveal.el\".")

(defcustom emacs-reveal-default-bibliography '("references.bib")
  "Default bibliography to assign to `org-ref-default-bibliography'.
A default helps to locate the bib file when the current buffer does not
specify one."
  :group 'org-export-emacs-reveal
  :type '(repeat :tag "List of BibTeX files" file)
  :package-version '(emacs-reveal . "7.1.0"))

(defcustom emacs-reveal-bibliography-entry-format
  '(("article" . "%a, %t, %j %v(%n), %p (%y). <a href=\"%U\">%U</a>")
    ("book" . "%a, %t, %u, %y. <a href=\"%U\">%U</a>")
    ("inproceedings" . "%a, %t, %b, %y. <a href=\"%U\">%U</a>")
    ("incollection" . "%a, %t, %b, %u, %y. <a href=\"%U\">%U</a>")
    ("misc" . "%a, %t, %i, %y.  <a href=\"%U\">%U</a>")
    ("phdthesis" . "%a, %t, %s, %y.  <a href=\"%U\">%U</a>")
    ("techreport" . "%a, %t, %i, %u (%y).  <a href=\"%U\">%U</a>")
    ("proceedings" . "%e, %t in %S, %u (%y)."))
  "Value to assign to `org-ref-bibliography-entry-format'.
This defines the layout of bibliography entries in presentations.
The default value displays article, book, inproceedings differently;
entries incollection, misc, and phdthesis are new, while entries
techreport and proceedings are defaults of `org-ref'."
  :group 'org-export-emacs-reveal
  :type '(alist :key-type (string) :value-type (string))
  :package-version '(emacs-reveal . "7.1.0"))

(defcustom emacs-reveal-cite-pkg 'org-re-reveal-ref
  "Default citation package to use.
If the value is `org-re-reveal-citeproc', upon loading, emacs-reveal adds
function `org-re-reveal-citeproc-filter-cite' to
`org-export-filter-paragraph-functions'."
  :group 'org-export-emacs-reveal
  :type '(choice (const org-re-reveal-ref)
                 (const org-re-reveal-citeproc))
  :package-version '(emacs-reveal . "9.0.0"))

(defvar emacs-reveal--cite-pkg-req
  (cond ((eq 'org-re-reveal-ref emacs-reveal-cite-pkg) 'org-ref)
        ((eq 'org-re-reveal-citeproc emacs-reveal-cite-pkg) 'citeproc))
  "Requirement for `emacs-reveal-cite-pkg'.")

(defun emacs-reveal--install-cite-pkg (explanation)
  "Show EXPLANATION and offer installation of `emacs-reveal--cite-pkg-req'."
  (unless (yes-or-no-p explanation)
    (error "Please install %s to use `emacs-reveal'" emacs-reveal--cite-pkg-req))
  (let ((package-archives (cons '("melpa" . "https://melpa.org/packages/") package-archives)))
    (package-refresh-contents)
    (package-install emacs-reveal--cite-pkg-req)
    (message-box
     "Installed %s.  Please restart Emacs to avoid issues with mixed Org installations."
     emacs-reveal--cite-pkg-req)))

;; org-ref and citeproc both have f as dependency.
;; If f is missing, offer to install emacs-reveal--cite-pkg-req.
;; Do not require emacs-reveal--cite-pkg-req here as that might pull in a
;; wrong Org version, since load-path has not been set up yet.
(package-initialize)
(condition-case nil
    (require 'f)
  (error
   (emacs-reveal--install-cite-pkg
    "Emacs-reveal: Dependency for citations not found.  Install from MELPA? ")))

(require 'f)
(defconst emacs-reveal-lisp-packages
  (list (f-join "org-mode" "lisp" "org-version.el")
        (f-join "org-re-reveal" "org-re-reveal.el")
        (f-join "org-re-reveal-citeproc" "org-re-reveal-citeproc.el")
        (f-join "org-re-reveal-ref" "org-re-reveal-ref.el")
        (f-join "oer-reveal" "oer-reveal.el"))
  "Lisp files of packages included as submodules.")

(defun emacs-reveal-submodules-ok ()
  "Check whether submodules are initialized properly.
Check whether (a) Lisp files for submodules in `emacs-reveal-lisp-packages'
are readable and (b) the JavaScript file \"reveal.js\" is readable.
If a check fails, return nil; otherwise, return directory of `emacs-reveal'."
  (let ((revealjs (f-join emacs-reveal-install-dir "emacs-reveal-submodules"
                          "reveal.js" "js" "reveal.js")))
    (when (and
           (not (memq nil
                      (mapcar #'file-readable-p
                              (mapcar (lambda (file)
                                        (f-join emacs-reveal-install-dir file))
                                      emacs-reveal-lisp-packages))))
           (file-readable-p revealjs))
      emacs-reveal-install-dir)))

(defun emacs-reveal-setup ()
  "Set up `emacs-reveal'.
If `emacs-reveal-managed-install-p' is t and a \".git\" is present,
invoke \"make setup\" to update `emacs-reveal' and submodules.
If submodules are present, add directories of Lisp packages to `load-path'."
  (when emacs-reveal-managed-install-p
    (if (file-readable-p (f-join emacs-reveal-install-dir ".git"))
        (let ((default-directory emacs-reveal-install-dir))
          (message "Emacs-reveal: \"make setup\" for submodules (large download upon first time; sets up Org mode) ...")
          (condition-case err
              (unless (= 0 (call-process "make" nil nil nil "setup"))
                (error "Status != 0"))
            (error (message-box "Emacs-reveal: Update with \"make setup\" failed: %s"
                                (error-message-string err))))
          (message "... done"))
      ;; Submodules might still be OK, e.g., in Docker.  Raise error if not.
      (unless (emacs-reveal-submodules-ok)
        (error "Must have a \".git\" subdirectory for managed install of `emacs-reveal'"))))
  (when (emacs-reveal-submodules-ok)
    (dolist (file emacs-reveal-lisp-packages)
      (add-to-list 'load-path (f-join emacs-reveal-install-dir
                                      (file-name-directory file))))))

(defun emacs-reveal-setup-oer-reveal ()
  "Set up `oer-reveal' for use with `emacs-reveal'.
If `oer-reveal' is used standalone, it manages installation and updating
of \"emacs-reveal-submodules\" itself.  When used as part of
`emacs-reveal' with properly installed submodules, `oer-reveal' should
not touch submodules.  Thus, set `oer-reveal-submodules-dir' to its
subdirectory under `emacs-reveal' and set
`oer-reveal-submodules-version' to nil.
Call `oer-reveal-setup-submodules', `oer-reveal-generate-include-files',
and `oer-reveal-publish-setq-defaults'."
  (when (emacs-reveal-submodules-ok)
    (let ((dir (f-join emacs-reveal-install-dir "emacs-reveal-submodules")))
      (setq oer-reveal-submodules-dir dir
            oer-reveal-submodules-version nil)))
  (oer-reveal-setup-submodules t)
  (oer-reveal-generate-include-files t)
  (oer-reveal-publish-setq-defaults))

;; Possibly update emacs-reveal (depending on emacs-reveal-managed-install-p);
;; set up load-path if necessary directories are present.
;; Afterwards, set up oer-reveal.
(emacs-reveal-setup)
(message "Using emacs-reveal with Org version: %s" (org-version))
(require 'oer-reveal-publish)
(emacs-reveal-setup-oer-reveal)

;; Set up bibliography in HTML.
(condition-case nil
    (require (eval 'emacs-reveal--cite-pkg-req))
  (error
   (emacs-reveal--install-cite-pkg
    (format
     "Emacs-reveal: Citation package %s not found.  Install from MELPA? "
     emacs-reveal--cite-pkg-req))))
(require (eval 'emacs-reveal-cite-pkg))
(cond ((eq 'org-re-reveal-ref emacs-reveal-cite-pkg)
       (setq org-ref-default-bibliography emacs-reveal-default-bibliography
             org-ref-bibliography-entry-format emacs-reveal-bibliography-entry-format))
      ((eq 'org-re-reveal-citeproc emacs-reveal-cite-pkg)
       (add-to-list 'org-export-filter-paragraph-functions
	            #'org-re-reveal-citeproc-filter-cite)))
(provide 'emacs-reveal)
;;; emacs-reveal.el ends here
